tower-http/src/classify_response.rs

```diff
@@ -0,0 +1,82 @@
+use http::{Response, StatusCode};
+use std::fmt;
+
+/// Classify requests results as either success or failure.
+///
+/// Not all services can rely entirely on the status code to determine if its a success or failure.
+/// gRPC for example has its own way of communicating that and always uses `200 OK`.
+pub trait ClassifyResponse<Body, Err> {
+    /// The type used to classify successful responses.
+    type OkClass;
+
+    /// The type used to classify failed responses.
+    type ErrClass;
+
+    /// Do the classification.
+    fn classify_request_result(
+        &self,
+        request_result: &Result<Response<Body>, Err>,
+    ) -> ResponseClassification<Self::OkClass, Self::ErrClass>;
+}
```

Comment:
I'm not sure if this API for response classification is really sufficient. 

For example, you mentioned gRPC in the docs. However, gRPC responses should be classified based on [the `grpc-status` trailer](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md), which is only sent _after_ a streaming body completes. Therefore, we can't classify gRPC responses based on the `http::Response` type --- we would need to wrap the body, returning a new response, and run the classification logic only when the body type's [`poll_trailers` method](https://docs.rs/http-body/0.4.0/http_body/trait.Body.html#tymethod.poll_trailers) returns trailers.

This is why linkerd's response-classification code has [two separate traits, `Classify` and `ClassifyEos`](https://github.com/linkerd/linkerd2-proxy/blob/e0271b4ee4841ef86fcbdb34f86b0fe725c502e2/linkerd/http-classify/src/lib.rs#L27-L53). A `Classify` can classify errors immediately, but in order to classify a response, it needs to produce an end-of-stream classifier that waits until the response's trailers are received. A design more like this is probably necessary in order to properly support gRPC...

Reply:
Uhh thats intersting. I didn't know gRPC could do that. Makes sense why this doesn't work then.

Seeing as you have lots of experience with this area would you perhaps want to make an initial version that you think would work? Could be as a PR to this PR or a separate new PR.

I'm not against copying what you have linkerd-proxy but my initial impression was that it was pretty complicated. Perhaps some docs and usage examples would help. Seems like the complexities are warranted though 😜

Reply:
> I'm not against copying what you have linkerd-proxy but my initial impression was that it was pretty complicated.

It is pretty complicated, but I think a lot of that complexity is kind of inherently necessary if you want to be able to include trailers in classification, which is kind of necessary to support gRPC. 

While we can't get away with not having it, I think it _might_ be possible to design an interface for `tower-http` where that additional complexity is _optional_...we might be able to have an interface where you can write a simple classifier based only on HTTP status codes without having to write two trait impls?

Reply:
I'm not really worried about things being complex if thats what it takes to also be correct. I think its about making "easy things easy and hard things possible". So ideally users who just want the defaults should be able to do `.layer(TraceLayer::new())` and move on. Whereas if deep customization requires implementing two traits then I don't think thats necessarily bad.

Reply:
Also, we use gRPC at Embark so I have some vested interested the gRPC classification working 😅